source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1314,17 +1384,63 @@ bool ServerContextImpl::isClientEcdsaCapable(const SSL_CLIENT_HELLO* ssl_client_
   return false;
 }
 
+OcspStapleAction ServerContextImpl::passesOcspPolicy(const ContextImpl::TlsContext& ctx) {
+  auto& response = ctx.ocsp_response_;
+  if (!Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.validate_ocsp_expiration_on_connection")) {
+    return response ? OcspStapleAction::Staple : OcspStapleAction::NoStaple;
+  }
+
+  if (ctx.is_must_staple_ && response && response->isExpired()) {
+    return OcspStapleAction::Fail;
+  }
+
+  if (response && !response->isExpired()) {
+    return OcspStapleAction::Staple;
+  }
+
+  // `response` is either not present or expired, and the action depends on the policy.
+  switch (ocsp_staple_policy_) {
+  case Ssl::ServerContextConfig::OcspStaplePolicy::SkipStaplingIfExpired:
+    return OcspStapleAction::NoStaple;
+  case Ssl::ServerContextConfig::OcspStaplePolicy::StaplingRequired:
+    return OcspStapleAction::Fail;
+  case Ssl::ServerContextConfig::OcspStaplePolicy::RejectConnectionOnExpired:
+    return response ? OcspStapleAction::Fail : OcspStapleAction::NoStaple;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
 enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
   const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
   // Fallback on first certificate.
   const TlsContext* selected_ctx = &tls_contexts_[0];
   for (const auto& ctx : tls_contexts_) {
-    if (client_ecdsa_capable == ctx.is_ecdsa_) {
+    if (client_ecdsa_capable == ctx.is_ecdsa_ && passesOcspPolicy(ctx) != OcspStapleAction::Fail) {
       selected_ctx = &ctx;
       break;
     }
   }
+
+  switch (passesOcspPolicy(*selected_ctx)) {
+  case OcspStapleAction::Staple: {
+    RELEASE_ASSERT(selected_ctx->ocsp_response_,
+                   "OCSP response must be present under OcspStapleAction::Staple");
+    auto& resp_bytes = selected_ctx->ocsp_response_->rawBytes();
+    RELEASE_ASSERT(
```

Comment:
style: I prefer the important stuff to not be inside an ASSERT (even if it's RELEASE_ASSERT). Capture the return value and RELEASE_ASSERT that it succeeded please.